pkg/mpc-examples/contracts/00-null-controller/NullControllerFactory.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// TODO license
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+// Core Infra
+import "./NullController.sol";
+import "../interfaces/IManagedPoolFactory.sol";
+import "@balancer-labs/v2-interfaces/contracts/vault/IVault.sol";
+import "@balancer-labs/v2-interfaces/contracts/pool-utils/IManagedPool.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/Create2.sol";
+
+/**
+ * @title NullControllerFactory
+ * @notice Factory for a Managed Pool and NullController.
+ * @dev Determines controller deployment address, deploys pool (w/ controller address as argument), then controller.
+ */
+contract NullControllerFactory {
+    mapping(address => bool) public isControllerFromFactory;
+
+    address public managedPoolFactory;
+    IVault public balancerVault;
+    bool public isDisabled;
+
+    uint256 private _nextControllerSalt;
+    address private _lastCreatedPool;
+
+    event ControllerCreated(address indexed controller, IVault vault, bytes32 poolId);
+
+    constructor(IVault vault, address factory) {
+        balancerVault = vault;
+        managedPoolFactory = factory;
+    }
+
+    /**
+     * @dev Return the address of the most recently created pool.
+     */
+    function getLastCreatedPool() external view returns (address) {
+        return _lastCreatedPool;
+    }
+
+    function create(IManagedPoolFactory.NewPoolParams memory params) external {
+        if (!isDisabled) {
+            bytes32 controllerSalt = bytes32(_nextControllerSalt);
+            _nextControllerSalt += 1;
+
+            bytes memory controllerCreationCode = abi.encodePacked(
+                type(NullController).creationCode,
+                abi.encode(balancerVault)
+            );
+            address expectedControllerAddress = Create2.computeAddress(
+                controllerSalt,
+                keccak256(controllerCreationCode)
+            );
+
+            // build arguments to deploy pool from factory
+            address[] memory assetManagers = new address[](params.tokens.length);
+            for (uint256 i = 0; i < assetManagers.length; i++) {
+                assetManagers[i] = expectedControllerAddress;
+            }
+
+            // TODO: instead of accepting the entirety of IManagedPoolFactory.NewPoolParams and
+            //       overwriting assetManagers and mustAllowlistLPs should we instead:
+            //          * extract all arguments and pass them individually?
+            //          * make a separate struct of the non-overwritten args, and use that to populate?
```

Comment:
I vote for the separate struct.

Reply:
added

Reply:
When populating `IManagedPoolFactory.NewPoolParams` from `MinimalPoolParams`, I'm doing it in order of the `IManagedPoolFactory.NewPoolParams` struct. This was initially to maintain parallel entries, but it does make it far less obvious what's being populated by the factory itself. Do you think this structure is good, or do you think it would be better to separate out the arguments that are not coming from the user?

Reply:
Current state:
```
IManagedPoolFactory.NewPoolParams memory fullParams;
fullParams.name = minimalParams.name;
fullParams.symbol = minimalParams.symbol;
fullParams.tokens = minimalParams.tokens;
fullParams.normalizedWeights = minimalParams.normalizedWeights;
fullParams.assetManagers = assetManagers;
fullParams.swapFeePercentage = minimalParams.swapFeePercentage;
fullParams.swapEnabledOnStart = minimalParams.swapEnabledOnStart;
fullParams.mustAllowlistLPs = false;
fullParams.managementAumFeePercentage = minimalParams.managementAumFeePercentage;
fullParams.aumFeeId = minimalParams.aumFeeId;
```
possible alternate way:
```
IManagedPoolFactory.NewPoolParams memory fullParams;
fullParams.name = minimalParams.name;
fullParams.symbol = minimalParams.symbol;
fullParams.tokens = minimalParams.tokens;
fullParams.normalizedWeights = minimalParams.normalizedWeights;
fullParams.swapFeePercentage = minimalParams.swapFeePercentage;
fullParams.swapEnabledOnStart = minimalParams.swapEnabledOnStart;
fullParams.managementAumFeePercentage = minimalParams.managementAumFeePercentage;
fullParams.aumFeeId = minimalParams.aumFeeId;

fullParams.assetManagers = assetManagers;
fullParams.mustAllowlistLPs = false;

```

Reply:
added